File: server/wave_api/compress.py
"""
DEFLATE payload compression for the LoRa-bound wire (Python side).

Mirror of Wave/src/sync/compress.ts. Both speak the self-describing zlib DEFLATE
format, so either side inflates the other's output regardless of the exact bytes
each library emits. Used by the envelope layer to compress the tokenized payload.
"""
from __future__ import annotations

import base64
import zlib


def deflate_to_b64(s: str) -> str:
    """UTF-8 string -> DEFLATE -> base64 ASCII."""
    return base64.b64encode(zlib.compress(s.encode("utf-8"))).decode("ascii")


def inflate_from_b64(b64: str) -> str:
    """base64 -> DEFLATE-inflate -> UTF-8 string."""
    return zlib.decompress(base64.b64decode(b64)).decode("utf-8")


def deflate_to_bytes(s: str) -> bytes:
    """Raw DEFLATE bytes — what the LoRa radio carries directly (no base64)."""
    return zlib.compress(s.encode("utf-8"))


def inflate_from_bytes(data: bytes) -> str:
    return zlib.decompress(data).decode("utf-8")

File: Wave/src/sync/compress.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * DEFLATE payload compression for the LoRa-bound wire.
 *
 * The codec turns a message into a tokenized array; this layer compresses that
 * array's JSON with zlib-format DEFLATE so the (often verbose) remediation/quiz
 * text rides as a fraction of the bytes. `server/wave_api/compress.py` is the
 * mirror: pako (here) and Python's `zlib` both speak the self-describing zlib
 * format, so either side inflates the other's output regardless of exact bytes.
 *
 * Synchronous on purpose — `buildEnvelope`/`Transport.publish` are sync.
 */
import pako from 'pako';

function u8ToB64(u8: Uint8Array): string {
  let bin = '';
  // Chunked to avoid call-stack limits on large payloads.
  const CH = 0x8000;
  for (let i = 0; i < u8.length; i += CH) {
    bin += String.fromCharCode(...u8.subarray(i, i + CH));
  }
  return btoa(bin);
}

function b64ToU8(b64: string): Uint8Array {
  const bin = atob(b64);
  const u8 = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) u8[i] = bin.charCodeAt(i);
  return u8;
}

/** UTF-8 string -> DEFLATE -> base64 (ASCII, safe for JSON/MQTT/REST today). */
export function deflateToB64(s: string): string {
  return u8ToB64(pako.deflate(s));
}

/** base64 -> DEFLATE-inflate -> UTF-8 string. */
export function inflateFromB64(b64: string): string {
  return pako.inflate(b64ToU8(b64), { to: 'string' });
}

/** Raw DEFLATE bytes — what the LoRa radio carries directly (no base64). */
export function deflateToBytes(s: string): Uint8Array {
  return pako.deflate(s);
}

export function inflateFromBytes(bytes: Uint8Array): string {
  return pako.inflate(bytes, { to: 'string' });
}
